lib/shared/drive-constraints.js

```diff
@@ -208,3 +208,162 @@ exports.isDriveValid = (drive, image) => {
+   */
+  OK: '',
+
+  /**
+   * @property {String} SIZE_NOT_RECOMMENDED
+   * @memberof COMPATIBILITY_STATUS_MESSAGES
+   *
+   * @description
+   * The image and drive compatibility is not recommended; happens when the
+   * image's recommended drive size doesn't match the actual drive size.
+   */
+  SIZE_NOT_RECOMMENDED: 'Not Recommended',
+
+  /**
+   * @property {String} TOO_SMALL
+   * @memberof COMPATIBILITY_STATUS_MESSAGES
+   *
+   * @description
+   * The drive is too small for the image.
+   */
+  TOO_SMALL: 'Too Small For Image',
+
+  /**
+   * @property {String} LOCKED
+   * @memberof COMPATIBILITY_STATUS_MESSAGES
+   *
+   * @description
+   * The drive is locked (SD cards) and cannot be written to.
+   */
+  LOCKED: 'Locked',
+
+  /**
+   * @property {String} SYSTEM
+   * @memberof COMPATIBILITY_STATUS_MESSAGES
+   *
+   * @description
+   * The drive is a system drive and should not be written to.
+   */
+  SYSTEM: 'System Drive',
+
+  /**
+   * @property {String} CONTAINS_IMAGE
+   * @memberof COMPATIBILITY_STATUS_MESSAGES
+   *
+   * @description
+   * The drive contains the image and therefore cannot be written to.
+   */
+  CONTAINS_IMAGE: 'Drive Contains Image'
+};
+
+/**
+ * @summary Drive/image compatibility status types.
+ * @public
+ * @type {Object}
+ *
+ * @description
+ * Status types classifying what kind of message it is, i.e. ok, error, warning.
+ */
+exports.COMPATIBILITY_STATUS_TYPES = {
+  OK: 0,
+  ERROR: 1,
+  WARNING: 2
+};
+
+/**
+ * @summary Get drive/image compatibility in an object
+ * @function
+ * @public
+ *
+ * @description
+ * Given an image and a drive, return their compatibility status object
+ * containing the status type (ERROR, WARNING, and OK), and accompanying
+ * status message.
+ *
+ * @param {Object} drive - drive
+ * @param {Object} image - image
+ * @returns {Object} compatibility status object
+ *
+ * @example
+ * const drive = {
+ *   device: '/dev/disk2',
+ *   name: 'My Drive',
+ *   size: 4000000000
+ * };
+ *
+ * const image = {
+ *   path: 'rpi.img',
+ *   size: 2000000000
+ *   recommendedDriveSize: 4000000000
+ * });
+ *
+ * const { type, message } = constraints.getDriveImageCompatibilityStatus(drive, image);
+ *
+ * if (type === constraints.COMPATIBILITY_STATUS_TYPES.WARNING) {
+ *   // do something
+ * } else if (type === constraints.COMPATIBILITY_STATUS_TYPES.ERROR) {
+ *   // do something else
+ * }
+ */
+exports.getDriveImageCompatibilityStatus = (drive, image) => {
+
+  // Mind the order of the if-statements if you modify.
+  if (exports.isSourceDrive(drive, image)) {
+    return {
+      type: exports.COMPATIBILITY_STATUS_TYPES.ERROR,
+      message: exports.COMPATIBILITY_STATUS_MESSAGES.CONTAINS_IMAGE
+    };
+  }
+
+  if (exports.isSystemDrive(drive)) {
+    return {
+      type: exports.COMPATIBILITY_STATUS_TYPES.WARNING,
+      message: exports.COMPATIBILITY_STATUS_MESSAGES.SYSTEM
+    };
+  }
+
+  if (!_.isNil(drive) && !exports.isDriveLargeEnough(drive, image)) {
```

Comment:
We should probably re-order these so that any ERROR conditions get tested before any WARNING conditions?
AIUI the code as it stands currently would mean that a drive that was both "system" and "locked" would only be labelled with "warning: system", whereas it probably makes sense to label it with "error: locked" ?